pkg/shell/manifests.ts

```diff
@@ -19,18 +19,146 @@
 
 import { JsonValue, JsonObject } from "cockpit";
 
+// Generic validation machinery
+
+class ValidationError extends Error { }
+
+const validation_path: string[] = [];
+
+function with_validation_path<T>(p: string, func: () => T): T {
+    validation_path.push(p);
+    try {
+        return func();
+    } finally {
+        validation_path.pop();
+    }
+}
+
+function validation_error(msg: string) {
+    console.error("Validation error:", JSON.stringify(validation_path), msg);
+}
+
+function import_string(val: JsonValue, fallback?: string): string {
+    if (typeof val == "string")
+        return val;
+    validation_error(`Not a string: ${JSON.stringify(val)}`);
+    if (fallback !== undefined)
+        return fallback;
+    throw new ValidationError();
```

Comment:
This is very close to what I had in mind, except that I'd handle "absent" (undefined) differently from "wrong type" -- after all, almost every field in a manifest is optional, and we shouldn't spew console errors for these. 

(and maybe `fallback` → `default`)

Reply:
Yes, there is some dicussion to be had. My idea here is that default values are not handled at all during validation. They are provided by the code that works with these data structures. The fallbacks here are only for the case were a value is of the wrong type, but we want to hobble along anyway.

The only fallbacks in use are:

 - something is supposed to be an array but is something else, then we continue with an empty array (after emitting the error)
 - something is supposed to be a dictionary but is something else, then we continue with an empty dict (after emitting an error)
 - when the "weight" of a keyword or the order of a menu entry is not a number, then we use 0 (after emitting an error)
 - when the "translate" of a keyword is not a boolean, then we continue with "false" (after emitting an error)

I think we could remove the last two cases actually. Using `{ weight: "heavy" }` as a keyword should probably disqualify (and skip) the whole entry.

Let's make a separate PR for this.

Reply:
We should think long and hard about how callers want to use it -- i.e. if a manifest override says e.g. `preload: 1` or doesn't specify preload, then it seems to me the only difference should be that the former causes a console warning. But in both cases the caller (specifying a reasonably default such as `[]` or `null`) would get that default in both cases?

Reply:
In other words, having two fallbacks (one for "wrong type" and another one for "absent") feels like accidental complexity and prone to confusion to me, WDYT?